### PR TITLE
[REVIEW] Remove deprecated `DataFrame.from_gpu_matrix`, `DataFrame.to_gpu_matrix`, `DataFrame.add_column` APIs and method parameters

### PR DESCRIPTION
**Pull request by galipremsagar**
- [x]  This PR removes deprecated methods: `DataFrame.from_gpu_matrix`, `DataFrame.to_gpu_matrix`, `DataFrame.add_column`
- [x]  Handles setting index correctly, and throwing an error while setting index instead of erroring when accessing later
 i.e.

```python
>>> import cudf
i>>> import cupy
>>> import numpy as np
>>> d_ary = cupy.empty((2, 3), dtype=np.int32)
>>> cudf.DataFrame(d_ary)
   0  1  2
0  0  0  0
1  0  0  0
>>> x = cudf.DataFrame(d_ary, index=['a'])     # This is incorrect, and we should error here
>>> x               # But we error when we try to access/print the dataframe.
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/core/dataframe.py", line 2494, in index
    raise ValueError(msg)
ValueError: Length mismatch: Expected axis has 2 elements, new values have 1 elements

# With this fix in this PR, we will now error during dataframe creation:
>>> x = cudf.DataFrame(d_ary, index=['a'])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/core/dataframe.py", line 255, in __init__
    self.index = new_df._index
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/core/dataframe.py", line 583, in __setattr__
    object.__setattr__(self, key, col)
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/core/dataframe.py", line 2494, in index
    raise ValueError(msg)
ValueError: Length mismatch: Expected axis has 2 elements, new values have 1 elements
```

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by beckernick** [comment]
@galipremsagar could you also update https://github.com/rapidsai/cudf/blob/branch-0.16/docs/cudf/source/10min-cudf-cupy.ipynb ? Looks like there's one use in there, too.

**Comment by galipremsagar** [comment]
> @galipremsagar could you also update https://github.com/rapidsai/cudf/blob/branch-0.16/docs/cudf/source/10min-cudf-cupy.ipynb ? Looks like there's one use in there, too.

Thanks Nick ! This should be updated now.

File: CHANGELOG.md [hunk 1]
```diff
@@ -118,6 +118,7 @@
 - PR #6409 Allow CuPy 8.x
 - PR #6407 Add RMM_LOGGING_LEVEL flag to Java docker build
 - PR #6438 Fetch nvcomp v1.1.0 for JNI build
+- PR 6442 Remove deprecated `DataFrame.from_gpu_matrix`, `DataFrame.to_gpu_matrix`, `DataFrame.add_column`, `dask_cudf.DataFrame.repartition` APIs and method parameters
```
Comment (kkraus14) [inline CHANGELOG.md:121]: Needs to be moved to 0.17 😄

File: docs/cudf/source/10min-cudf-cupy.ipynb [hunk 2]
```diff
@@ -62,15 +45,15 @@
...
+      "CPU times: user 346 µs, sys: 400 µs, total: 746 µs\n",
+      "Wall time: 646 µs\n"
      ]
     },
     {
      "name": "stderr",
      "output_type": "stream",
      "text": [
-      "/conda/envs/cudf/lib/python3.7/site-packages/cudf/io/dlpack.py:74: UserWarning: WARNING: cuDF to_dlpack() produces column-major (Fortran order) output. If the output tensor needs to be row major, transpose the output of this function.\n",
+      "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/io/dlpack.py:74: UserWarning: WARNING: cuDF to_dlpack() produces column-major (Fortran order) output. If the output tensor needs to be row major, transpose the output of this function.\n",
```
Comment (kkraus14) [inline docs/cudf/source/10min-cudf-cupy.ipynb:56]: We should fix this path to use the container paths as they're a bit cleaner.

File: docs/cudf/source/10min-cudf-cupy.ipynb [hunk 3]
```diff
@@ -309,7 +292,7 @@
      "name": "stderr",
      "output_type": "stream",
      "text": [
-      "/conda/envs/cudf/lib/python3.7/site-packages/cudf/io/dlpack.py:33: UserWarning: WARNING: cuDF from_dlpack() assumes column-major (Fortran order) input. If the input tensor is row-major, transpose it before passing it to this function.\n",
+      "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/io/dlpack.py:33: UserWarning: WARNING: cuDF from_dlpack() assumes column-major (Fortran order) input. If the input tensor is row-major, transpose it before passing it to this function.\n",
```
Comment (kkraus14) [inline docs/cudf/source/10min-cudf-cupy.ipynb:295]: here as well

File: docs/cudf/source/10min-cudf-cupy.ipynb [hunk 4]
```diff
@@ -975,6 +958,14 @@
    "execution_count": 13,
    "metadata": {},
    "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/io/dlpack.py:74: UserWarning: WARNING: cuDF to_dlpack() produces column-major (Fortran order) output. If the output tensor needs to be row major, transpose the output of this function.\n",
```
Comment (kkraus14) [inline docs/cudf/source/10min-cudf-cupy.ipynb:965]: +1

File: docs/cudf/source/10min-cudf-cupy.ipynb [hunk 5]
```diff
@@ -1255,10 +1246,18 @@
    "execution_count": 17,
    "metadata": {},
    "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/cudf/io/dlpack.py:74: UserWarning: WARNING: cuDF to_dlpack() produces column-major (Fortran order) output. If the output tensor needs to be row major, transpose the output of this function.\n",
```
Comment (kkraus14) [inline docs/cudf/source/10min-cudf-cupy.ipynb:1253]: +1

File: python/dask_cudf/dask_cudf/core.py [hunk 6]
```diff
@@ -317,27 +317,6 @@ def var(
...
-            )
-            warnings.warn(
-                "Rearranging data by column hash. Divisions will lost. "
-                "Set ignore_index=False to preserve Index values."
-            )
-            ignore_index = kwargs.pop("ignore_index", True)
-            return self.shuffle(
-                on=columns, ignore_index=ignore_index, **kwargs
-            )
-        return super().repartition(*args, **kwargs)
```
Comment (kkraus14) [inline python/dask_cudf/dask_cudf/core.py:339]: Why was this removed?

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: python/dask_cudf/dask_cudf/core.py (hunk 6 above)
Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/core.py:339]: The only reason this seems to be existing is to check if `columns` exists. Now that we are removing the deprecated params from this function we'll be left with call to super call function(`return super().repartition(*args, **kwargs)`). Hence removed this function

File: CHANGELOG.md (hunk 1 above)
Comment (galipremsagar) [reply CHANGELOG.md]: Missed it 🤣  Thanks

File: python/dask_cudf/dask_cudf/core.py (hunk 6 above)
Comment (kkraus14) [reply python/dask_cudf/dask_cudf/core.py:339]: cc @rjzamora

Can you confirm this is a deprecated function that we're safe to remove?

Comment (rjzamora) [reply python/dask_cudf/dask_cudf/core.py:339]: > Can you confirm this is a deprecated function that we're safe to remove?

Yes - This can be removed as long as it is no longer being used in tpcxbb (which I doubt it is) cc @beckernick

Comment (beckernick) [reply python/dask_cudf/dask_cudf/core.py:339]: We still `df.repartition(columns=col)` quite a bit. If this functionality is fully covered by the Dask `df.shuffle` API, we should be fine to remove it. I'll file an issue to switch to the shuffle API

Comment (rjzamora) [reply python/dask_cudf/dask_cudf/core.py:339]: > We still df.repartition(columns=col) quite a bit. If this functionality is fully covered by the Dask df.shuffle API, we should be fine to remove it. I'll file an issue to switch to the shuffle API on TPCxBB

Okay - Glad I checked with you :)

The functionality is indeed the same as `shuffle` (where the API is consistent with `dask.dataframe`)

Comment (kkraus14) [reply python/dask_cudf/dask_cudf/core.py:339]: @galipremsagar lets defer removing this to a follow up, maybe just add a #TODO

Comment (galipremsagar) [reply python/dask_cudf/dask_cudf/core.py]: @kkraus14 Done

**Review by kkraus14 — APPROVED** [review]

**Review by shwina — APPROVED** [review]
lgtm!

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests
